### ISSUE TITLE
[no-unnecessary-type-assertion] False positives with `!` assertions

### ISSUE THREAD

**Issue by samking** [typescript-eslint/typescript-eslint#1618]
**Repro**

test.ts
```TS
export function test() {                                                            
  let numOrNull: number | null = null                                               
  function needsANumber(num: number) {                                              
    console.log(num)                                                                
  }                                                                                 
  needsANumber(numOrNull!)                                                          
}                                                                                   
```

.eslintrc.js
```JS
module.exports = {                                                                  
  settings: {                                                                       
    'import/parsers': {'@typescript-eslint/parser': ['.ts', '.tsx']},               
  },                                                                                
  parser: '@typescript-eslint/parser',                                              
  plugins: ['@typescript-eslint'],                                                  
  rules: {                                                                          
    '@typescript-eslint/no-unnecessary-type-assertion': 'error',                    
  },                                                                                
}                                                                                   
```

**Expected Result**

There is no lint failure.  Specifically, the `!` in `needsANumber(numOrNull!)` is necessary to change `numOrNull` from type `number | null` to type `number`, so it should not give a lint error.

Removing the `!` causes the Typescript compiler to give an error, as expected, so my only recourse is to ignore the lint error.

**Actual Result**

There is a lint error: `error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion`

**Additional Details**

Originally, I ran into this issue with a templated value (possibly related to https://github.com/typescript-eslint/typescript-eslint/issues/1410) in React.  E.g., `const [foo, setFoo] = useState<number | null>(null)`.  However, since the problem seems to come up with any possibly-null type rather than just templated values, I thought I'd file a separate issue.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.20.0` |
| `@typescript-eslint/parser`        | `2.20.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.7.0` |

**Comment by bradzacher** [comment]
Do you have `strictNullChecks` turned on?
If not, then this rule will be pretty flakey on you, esp with the non-null assertion operator.

**Comment by samking** [comment]
> Do you have strictNullChecks turned on?

In my `tsconfig.json`, I have `strict` set to true, so my compiler is using strict null checks.  Is that what you were asking, or do I separately also need to tell typescript-eslint to use strict mode?

**Comment by bradzacher** [comment]
Good to hear that fixed it, I'll go ahead and close this then.

Note you can use `project: ['./tsconfig.json', './packages/project1/tsconfig.json', ...]` to specify multiple project configs if you have them, or you can use globs as well:
`project: ['./tsconfig.eslint.json', './packages/*/tsconfig.json'],`

https://github.com/typescript-eslint/typescript-eslint/blob/7452e7d4f53da73fc1b15ee5bb8434cd9219676f/.eslintrc.js#L134

**Comment by samking** [comment]
Thanks!
